### PR TITLE
Optimized GenTemp by a factor of 10x

### PR DESCRIPTION
**Pull request by kbatbouta**
Changes:

- ~Now GenTemp patch works by checking a cache dict for cells occupied by fridges thus no looping over things in a cell.~
- Now the mod disable rotting rather than changing temperature. However the only side effect is the "frozen" label won't show up (Stay empty).
- Added FridgeCache used to store fridge things cells.

It's about ~2.5x to 4x~ 10x faster then before.

### CODE REVIEW

File: Source/FridgeCache.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+namespace RimFridge
+{
+    [StaticConstructorOnStartup]
+    public static class FridgeCache
+    {
+        public static Dictionary<IntVec3, CompRefrigerator>[] FridgeGrid = new Dictionary<IntVec3, CompRefrigerator>[10];
+
+        static FridgeCache()
+        {
+            for (int i = 0; i < 10; i++) FridgeGrid[i] = new Dictionary<IntVec3, CompRefrigerator>(200);
```
Comment (KiameV) [inline Source/FridgeCache.cs]: This seems like an awful lot of pre-allocated space. Why not:
`List<Dictionary<IntVec3, CompRefrigerator> FridgeGrid  = List<Dictionary<IntVec3, CompRefrigerator>()`?
Currently this is allocating 128 kb of memory for nothing.

Comment (KiameV) [reply Source/FridgeCache.cs]: I've also not done much with map indexes, might it be better to do:
Dictionary<int, <Dictionary<IntVec3, CompRefrigerator>> FridgeGrid = Dictionary<int, <Dictionary<IntVec3, CompRefrigerator>>()?

File: Source/CompRefrigerator.cs [hunk 2]
```diff
@@ -169,6 +169,13 @@ public override void CompTickRare()
             {
                 powerTrader.PowerOutput = -((CompProperties_Power)powerTrader.props).basePowerConsumption * ((powerMultiplier * 0.9f) + 0.1f);
             }
+
+            //FridgeCache.FridgeGrid
+            var index = map.Index;
+            foreach (IntVec3 cell in GenAdj.OccupiedRect(parent))
+            {
+                FridgeCache.FridgeGrid[index][cell] = this;
```
Comment (KiameV) [inline Source/CompRefrigerator.cs]: I'd suggest making this a method in FridgeCache:
```
public void AddFridgeToGrid(Map map, IntVec3 cell)
{
    if (!FridgeCache.TryGetValue(map.index, out Dictionary<IntVec3, CompRefrigerator> d))
    {
        d = new Dictionary<IntVec3, CompRefrigerator>(200);
        FridgeCache.Add(map.Index, d);
    }
    d.Add(cell);
}
```

File: Source/FridgeCache.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        }
+    }
+
```
Comment (KiameV) [inline Source/FridgeCache.cs]: What about a patch for when a Map is deleted to clear the cache of the old map?

**Review by KiameV — CHANGES_REQUESTED** [review]

File: Source/FridgeCache.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+﻿using Verse;
+using System.Collections.Generic;
+using HarmonyLib;
+
+namespace RimFridge
+{
+    [StaticConstructorOnStartup]
+    public static class FridgeCache
+    {
+        public static Dictionary<IntVec3, CompRefrigerator>[] FridgeGrid = new Dictionary<IntVec3, CompRefrigerator>[10];
```
Comment (KiameV) [inline Source/FridgeCache.cs:10]: Are we guaranteed to have a maximum of 10 maps?
This really needs to be a List if not a Dictionary

File: Source/RimFridge.csproj [hunk 5]
```diff
@@ -1,5 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (KiameV) [inline Source/RimFridge.csproj]: Please remove this file from the PR

File: Source/FridgeCache.cs (hunk 4 above)
Comment (kbatbouta) [reply Source/FridgeCache.cs:10]: Yes the game have a hard limit of 5 maps (colonies), and i don't think many users use more 10 maps at a time.

File: Source/RimFridge.csproj (hunk 5 above)
Comment (kbatbouta) [reply Source/RimFridge.csproj]: OK

File: Source/FridgeCache.cs (hunk 3 above)
Comment (kbatbouta) [reply Source/FridgeCache.cs]: GC for this isn't a big deal if i'm honest since maps are rarely created or deleted.

Comment (kbatbouta) [reply Source/FridgeCache.cs]: GC can be an issue for other stuff that is created a lot more often.

File: Source/FridgeCache.cs (hunk 1 above)
Comment (kbatbouta) [reply Source/FridgeCache.cs]: No this will destroy the fast access time.

Comment (kbatbouta) [reply Source/FridgeCache.cs]: Dict are to be avoided when arrays are usable.

File: Source/RimFridge.csproj (hunk 5 above)
Comment (kbatbouta) [reply Source/RimFridge.csproj]: However remember to add the `FridgeCache.cs` back to the project.

File: Source/FridgeCache.cs (hunk 1 above)
Comment (KiameV) [reply Source/FridgeCache.cs]: The map index cannot be guaranteed so a growable solution is required. Dictionary is the easier route. If you wanted to you can create a new container type the wraps a list and grows the list as needed as the index increases.

File: Source/FridgeCache.cs (hunk 4 above)
Comment (KiameV) [reply Source/FridgeCache.cs:10]: I'm not having a staticly sized array in code. I've seen this go bad too many times in my life.

File: Source/FridgeCache.cs (hunk 1 above)
Comment (KiameV) [reply Source/FridgeCache.cs]: Remember a Dictionary in C# is a HashMap. For well distributed hashes a hash map has constant lookup time which is the same as an array (i will admit on the cpu there's additional computations needed for the modulo (in this case against an integer so very fast) and possible collisions (though the number of fridges will be small and will be well distributed)).

**Comment by KiameV** [comment]
I've decided to take this idea and instead of keeping track of which maps exist and instead make FridgeCache extend MapComponent.
I've removed the patch for CompRottable since it was breaking the labeling of the food and instead patch Thing.AmbientTemperature

**Comment by kbatbouta** [comment]
The whole idea was to avoid `Thing.AmbientTemperature` because it is so slow.

**Comment by KiameV** [comment]
Oh? In my testing the change from the original patch to patching Thing.AmbientTemperature  went from 3 ms to .1 ms. How is Thing.AmbientTemperature slow?

**Comment by KiameV** [comment]
There are also ways around slow methods. I tested doing a Prefix on AmbientTemperature vs Postfix. The time went from .11 ms with the postfix to .075 ms with the prefix. 0.04 ms. To me, if RW community views 0.04 ms as "slow" then this community really needs to figure out better things to do.

In any case I've gone ahead and change RimFridge to prefix AmbientTemperature with a high priority with the condition that it only eats the request when there is a (non-pawn | dead pawn) on a CompRefridge that's active.

**Comment by kbatbouta** [comment]
My goal is to avoid the slow temperature system. If you pulled it off then
good for you. However attacking me won't change the facts and i would like
to ask if your testing was performed on 3x.

**Comment by kbatbouta** [comment]
I apologize i confused it with the old system that use the temperature.

**Comment by KiameV** [comment]
As Dubwise says on the github page, "Don't go hunting down modders and demanding they fix things using this as evidence. The logging may not be accurate and may not tell the complete story as something else could cause someones mod to run slowly."

I can become annoyed when pestered over small issues after taking clear measurements and taking into account all other mods that either tie into/use this mod and its users. Performance analyzers are powerful tools and needs to be used with caution.

Thank you for bringing this issue to my attention and coming up with a PoC for making it better.

**Comment by kbatbouta** [comment]
First, i would like to note that i didn't go hunting down anyone. I do this for a living and i had assumed you would be more accepting of what i had to say and what i had to show. I came here with one possible solution, you found an other, great! the more angles to attack the problem the better. I didn't insult you or hunt you down. My statement about the temperature system is a fact measured without the analyzer (rather using DotTrace) and am disappointed that you would assume that. As i didn't make any assumption about you and even i acknowledge that i made a mistake.

**Pull request closed without merging**